src/main/java/spoon/reflect/visitor/CtAbstractBiScanner.java

```diff
@@ -19,81 +19,15 @@
 import spoon.reflect.declaration.CtElement;
 import spoon.reflect.path.CtRole;
 
-import java.util.ArrayDeque;
-import java.util.Collection;
-import java.util.Deque;
-import java.util.Iterator;
-
 /**
- * This abstract bi scanner class declares all scan methods useful for CtBiScannerDefault
+ * Defines the core bi-scan responsibility.
  */
 public abstract class CtAbstractBiScanner extends CtAbstractVisitor {
-	protected Deque<CtElement> stack = new ArrayDeque<>();
-
-	protected void enter(CtElement e) {
-	}
-
-	protected void exit(CtElement e) {
-	}
-
-	protected boolean isNotEqual = false;
-
-	public boolean biScan(Collection<? extends CtElement> elements, Collection<? extends CtElement> others) {
-		return biScan(null,  elements, others);
-	}
-	public boolean biScan(CtRole role, Collection<? extends CtElement> elements, Collection<? extends CtElement> others) {
-		if (isNotEqual) {
-			return isNotEqual;
-		}
-		if (elements == null) {
-			if (others != null) {
-				return fail();
-			}
-			return isNotEqual;
-		} else if (others == null) {
-			return fail();
-		}
-		if ((elements.size()) != (others.size())) {
-			return fail();
-		}
-		for (Iterator<? extends CtElement> firstIt = elements.iterator(), secondIt = others.iterator(); (firstIt.hasNext()) && (secondIt.hasNext());) {
-			biScan(role, firstIt.next(), secondIt.next());
-		}
-		return isNotEqual;
-	}
 
-	public boolean biScan(CtRole role, CtElement element, CtElement other) {
-		return biScan(element, other);
-	}
-	public boolean biScan(CtElement element, CtElement other) {
-		if (isNotEqual) {
-			return isNotEqual;
-		}
-		if (element == null) {
-			if (other != null) {
-				return fail();
-			}
-			return isNotEqual;
-		} else if (other == null) {
-			return fail();
-		}
-		if (element == other) {
-			return isNotEqual;
-		}
+	/** This method is called to compare `element` and `other` when traversing two trees in parallel.*/
+	public abstract boolean biScan(CtElement element, CtElement other);
```

Comment:
CtAbstractBiScanner sounds now like general purpose bi-scanner. Do I understood it well? If yes, then please change comments here (it is no more about "compare") and change return value from `boolean` to `void`
WDYT?

Reply:
> CtAbstractBiScanner sounds now like general purpose bi-scanner. Do I understood it well?

yes, correct.

I agree with you, done.